src/transport/raw/PeerTCPParamsStorage.cpp

```diff
@@ -0,0 +1,339 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the Peer TCP Parameter Storage object that provides
+ *      APIs for storage and retireval of the TCP parameters for a given peer
+ *      node scoped to a fabric.
+ */
+
+#include <transport/raw/PeerTCPParamsStorage.h>
+
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+
+namespace chip {
+namespace Transport {
+
+constexpr TLV::Tag PeerTCPParamsStorage::kSupportedTransportsTag;
+constexpr TLV::Tag PeerTCPParamsStorage::kMaxTCPMessageSizeTag;
+
+StorageKeyName PeerTCPParamsStorage::GetStorageKey(const ScopedNodeId & node)
+{
+    return DefaultStorageKeyAllocator::TCPPeerParams(node.GetFabricIndex(), node.GetNodeId());
+}
+
+CHIP_ERROR PeerTCPParamsStorage::FindByScopedNodeId(const ScopedNodeId & node, uint16_t & supportedTransports,
+                                                    uint32_t & maxTCPMessageSize)
+{
+    ReturnErrorOnFailure(LoadTCPParamsFromStorage(node, supportedTransports, maxTCPMessageSize));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR PeerTCPParamsStorage::SaveTrackedNodesList(const TrackedNodesList & list)
+{
+    std::array<uint8_t, MaxNodesListSize()> buf;
```

Comment:
how much is this MaxNodesListSize?

Reply:
Is the stack ok? How about a ScopedMemoryBuffer so we use heap and tools will detect out of bounds if it applies? We seem to assume large systems here anyway. Large buffers on the stack seem worse than large buffers in the heap.

Reply:
What happens if this size is exceeded?  What happens if a client needs to speak to more than 32 different nodes?

Reply:
One of the nodes(index 0) gets replaced. This is also a compile-time configurable value in TCPConfig.h.

Reply:
> One of the nodes(index 0) gets replaced

I guess the real question is: if it's totally ok to lose track of this info for nodes, why are we keeping track of it at all?

Reply:
Keeping track of the nodes list helps with management of the stored info.
The nodeslist has a finite size and will eventually fill up if there are more nodes that are being interacted with. But if an entry is replaced by a new one when the list is full, it can always be inserted back when a CASE negotiation happens again with that node.

Reply:
That's not what I asked.  What I asked is: why are we doing _any_ of the work this PR does if it's totally ok to just lose the stored information?

Reply:
Losing the information is not ok, but only an outcome of limited resources. The list being finite, the node could reach a point where it would lose the info for some peer node if the latter was replaced. But at the same time, there is the underlying premise that nodes would establish CASE over MRP for other various operations(subscriptions, events, time sync, commands whatever) from time to time, and the node info would get replenished passively. We can optimize further by implementing a better cache replacement strategy(LRU, etc). Also, increasing the size of the nodes list to a larger value, say 64 or 128 might practically eliminate the replacement loss. Resource-wise, this should be ok on a device that supports TCP.